src/controllers/colormapperjsproxy.cpp

```diff
@@ -0,0 +1,46 @@
+#include <QScriptValueIterator>
+
+#include "controllers/colormapperjsproxy.h"
+
+ColorMapperJSProxy::ColorMapperJSProxy(QScriptEngine* pScriptEngine, QMap<QRgb, QVariant> availableColors)
+    : m_pScriptEngine(pScriptEngine) {
+        m_colorMapper = new ColorMapper(availableColors);
+}
+
+QScriptValue ColorMapperJSProxy::getNearestColor(uint colorCode) {
+    QPair<QRgb, QVariant> result = m_colorMapper->getNearestColor(static_cast<QRgb>(colorCode));
+    QScriptValue jsColor = m_pScriptEngine->newObject();
+    jsColor.setProperty("red", qRed(result.first));
+    jsColor.setProperty("green", qGreen(result.first));
+    jsColor.setProperty("blue", qBlue(result.first));
+    return jsColor;
+}
+
+QScriptValue ColorMapperJSProxy::getNearestValue(uint colorCode) {
+    QPair<QRgb, QVariant> result = m_colorMapper->getNearestColor(static_cast<QRgb>(colorCode));
+    return m_pScriptEngine->toScriptValue(result.second);
+}
+
+QScriptValue ColorMapperJSProxyConstructor(QScriptContext* pScriptContext, QScriptEngine* pScriptEngine) {
+    QMap<QRgb, QVariant> availableColors;
+    DEBUG_ASSERT(pScriptContext->argumentCount() == 1);
```

Comment:
this code has a lot of debug asserts that feel like either they will never happen or they need to be properly handled.  It feels like you are putting these in because you're not sure if everything works right.  debug asserts are not a replacement for writing unit tests, and if you want to confirm that everything is working correctly then we need more tests.

Reply:
Assert for the impossible. If something is never expected to happen then a debug assert is the correct choice.

Debug assertions and unit tests serve a different purpose. In unit tests you test for what you expect, whereas debug assertions are guard rails that test for the unexpected. We should not trade one for the other.

Reply:
Yeah, I'm wondering if asserts are correct here. AFAIK asserts are used to catch *programming errors*, not user data checks. Is a controller script considered user data or code?

For example, this assertion would be triggered by a controller script that instantiates a `ColorMapper` object like this:

```js
var mapper = ColorMapper();
```

Reply:
If these debug asserts can be tripped by controller scripts, then at the very least there should be errors logged to help the user understand the assertion.  VERIFY_OR_DEBUG_ASSERT might be a better fit

Reply:
The best would be to throw a script error that pops up (like when your JS is broken), but I don't know how.

Reply:
controllerengine.cpp seems to do that, and it's based on qscriptengine's error catching: https://doc.qt.io/archives/qt-4.8/qscriptengine.html#uncaughtException

Maybe your code can throw an error in js such that this code catches it.

Reply:
Done.